### PR TITLE
!str #16992 Rework Source and Sink name parameter

### PR DESCRIPTION
**Pull request by patriknw**
I don't like this, but given that we really need the name parameter for the naming of the shape I found no other way.

`onComplete` javadsl signature was wrong

### CODE REVIEW

**Comment by rkuhn** [comment]
Just skimmed it for now: did you check that there is no method without parens anymore? Unfortunately this seems to make a difference when it comes to inferred type arguments and overloading, so we should do that consistently as well. An example of what I mean is that `.runWith(Sink.head)` does not work anymore without parens.

**Comment by patriknw** [comment]
@rkuhn `runWith(Sink.head)` actually works if I declare it as this:

    def head[T]: Sink[T, Future[T]]
    def head[T](name: String): Sink[T, Future[T]]

but then `runWith(Sink.head())` doesn't work

Otherwise we can declare it as 

    def head[T](name: String = ""): Sink[T, Future[T]]

and then only `runWith(Sink.head())` will work, and the parens will also be required for `val futureSink = Sink.head[Seq[Int]]()`

I have already used the default parameter value approach at places where there are more parameters.

WDYT?

**Comment by bantonsson** [comment]
I think that we should go with `Sink.head()`(and the same for cancelled), or if we don't then all other with `()` at the end (like ignore) should drop theirs as well.

**Comment by rkuhn** [comment]
I’d recommend to stay away from default arguments as they are not extensible in a binary compatible fashion, and source compatibility can also be challenging. So, thanks for your research, I think using overloaded methods with at least one argument list each is the way to go.

**Comment by patriknw** [comment]
The drawback with the overload is the copies of the docs.

but if we use overloads anyway, why not use the variant without argument list to support `Sink.head`?

**Comment by rkuhn** [comment]
If that works in all cases, then yes. I ran into some issue that might well have involved another bug on my part. The problem is if Scalac can’t figure that one out there is nothing you can write to disambiguate—parens are then not allowed.

**Comment by patriknw** [comment]
ok, I'll try it in large scale

**Comment by patriknw** [comment]
@rkuhn one surprise when I removed default arguments.
Changing:

    def foreach[T](f: T ⇒ Unit, name: String = ""): Sink[T, Future[Unit]]

to:

    def foreach[T](f: T ⇒ Unit, name: String): Sink[T, Future[Unit]]
    def foreach[T](f: T ⇒ Unit): Sink[T, Future[Unit]]

Then compilation error of for example:

    val echoHandler = Sink.foreach[StreamTcp.IncomingConnection] { _.flow.join(Flow[ByteString]).run() }


    missing parameter type for expanded function ((x$9) ⇒ x$9.flow.join(Flow[ByteString]).run())

**Comment by rkuhn** [comment]
Hmm, yes, that only works with default arguments … aargh. Overloading is evil, all kinds of it. Maybe a better approach is to add a `.named` method on Sink and thereby completely factor out the naming aspect?

**Comment by patriknw** [comment]
I can interpret `.named` in several ways, one is `Sink.named("blah").head`

What is the name used for in the shape? Do we really need to make that configurable? Is it possible to grab it afterwards from the `OperationAttribtues`?

**Comment by rkuhn** [comment]
The point is to _set_ the OperationAttributes of this (Sink) stage. I was thinking more like
~~~scala
Sink.head[Int].named("TheHead")
~~~

**Comment by rkuhn** [comment]
To add more color: the inlet of this sink should then have `"TheHead"` in its name, so that when you forget to wire it up you’ll get a nice and explanative error message.

**Comment by patriknw** [comment]
that sounds more in line with the existing `Sink.head[Int].withAttributes`, I will give it a try

**Comment by patriknw** [comment]
@rkuhn I have reworked this completely according to our discussion. Much better.

* Remove name parameter (no overloads), naming is performed using `.withAttributes` or the new
  convenience `.named`. Those adds the OperationAttribute.Name and also change the name of the
  shape Inlet and Outlet.

* Remove Source/Sink parameter list for 0 parameter methods,
  this allows usage of `Sink.head` instead of `Sink.head()`

**Comment by bantonsson** [comment]
LGTM

One question though. What name will an already named source (like `IterableSource`) have if you call `named(...)` on it? are the operational attributes overriden?

**Comment by patriknw** [comment]
I assume last wins. We have an additional ticket for revisiting the actual naming: #15707

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 1]
```diff
@@ -31,15 +34,24 @@ abstract class SinkModule[-In, Mat](val shape: SinkShape[In]) extends Module {
   }
 
   override def subModules: Set[Module] = Set.empty
+
+  def amendShape(attr: OperationAttributes): SinkShape[In] = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: why is this needed to be public?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: stupid question, this is the Module.

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: well, it could be restricted to `protected`

**Comment by drewhk** [comment]
It is not exactly last wins, it is hierarchic. So if you name an iterable it will have three names, the user given, inside that the name "iterable" and inside that the names "emptySource" and "stageFactory". How we calculate a display name from that is up to us. When I implemented toString on Source/Sink/Flow they always used the outermost name. A deepToString would actually show the innermost names and the hierarchic nesting (maybe showing the name hierarchy, too). When we name actors we take the whole sequence AFAIK ("myName-iterable-stageFactory").

**Comment by patriknw** [comment]
Refs #16992
